game-app/game-core/src/main/java/games/strategy/engine/data/DefaultAttachment.java

```diff
@@ -156,4 +151,69 @@ public final boolean equals(final Object obj) {
             Objects.toString(attachedTo, null), Objects.toString(other.attachedTo, null))
         && (Objects.equals(name, other.name) || this.toString().equals(other.toString()));
   }
+
+  protected Territory getTerritoryOrThrow(String name) throws GameParseException {
+    return Optional.ofNullable(getData().getMap().getTerritory(name))
+        .orElseThrow(() -> new GameParseException("No territory named: " + name + thisErrorMsg()));
+  }
+
+  protected List<GamePlayer> parsePlayerList(final String value, List<GamePlayer> existingList)
+      throws GameParseException {
+    for (final String name : splitOnColon(value)) {
+      if (existingList == null) {
+        existingList = new ArrayList<>();
+      }
```

Comment:
I dislike this kind of if check. If the parameter is passed we modify it and return it, otherwise we return a new object and return that. I would appreciate if this method always returns a new list and  the callers call either addAll on an existing collection or assign the new collection to the result. I think this would be cleaner code style. Applies to the other variants of this down below too of course

Reply:
I agree it's messy, but the point of this helper is to reduce duplicate code. Let's say it were changed to:

```
  protected List<GamePlayer> parsePlayerList(final String value)
      throws GameParseException {
    List<GamePlayer> players = new ArrayList<>();
    for (final String name : splitOnColon(value)) {
      existingList.add(getPlayerOrThrow(name));
    }
    return players;
  }
```

Then, what happens to the callers (we have 12 of them)?

Something like this:

```
  private void setHelpPayTechCost(final String value) throws GameParseException {
    helpPayTechCost = parsePlayerList(value, helpPayTechCost);
  }
```

Becomes:

```
  private void setHelpPayTechCost(final String value) throws GameParseException {
    if (helpPayTechCost == null) {
      helpPayTechCost = new ArrayList<>();
    }
    helpPayTechCost.addAll(parsePlayerList(value));
  }
```

I'd guess that codeclimate would flag it as duplicated code.

There are other ways to formulate it, but all are messy. For example, we could do:

```
helpPayTechCost = Optional.ofNullable(helpPayTechCost).orElse(new ArrayList<>());
helpPayTechCost.addAll(parsePlayerList(value));
```

Or we could do:

```
helpPayTechCost = createIfNull(helpPayTechCost);
helpPayTechCost.addAll(parsePlayerList(value));
```

Or:
```
helpPayTechCost = mergeLists(helpPayTechCost, parsePlayerList(value));
```

with the latter basically wrapping one of the above versions.

They're all kind of messy and the current approach is the one that results in the least code at call sites.

Reply:
I'll keep as-is since it's the most concise, but I'm also OK with something like the `mergeLists()` version. We can do that as a follow up later if we like.